### DISCUSSION TITLE
Implementing a set of boundary conditions

### DISCUSSION THREAD

**Discussion by Meghasemi-96** [idaholab/moose#25296] [Q&A General]
Hi, 
Hope you are doing well.

I have a question regarding applying boundary conditions. I am dealing with two variables, namely `p1` and `p2` (pressure for Fick's law). When it comes to the inlet boundary condition, I have the equality of p1=p2=4, and the values are already set accordingly with DirichletBC. On the other hand, for the outlet, two different conditions apply. Firstly, `p1+p2=1` , and I have the condition `J1/J2=p1/p2`, indicating that the ratio of fluxes is equal to the ratio of pressures. It's important to note that `p1` and `p2` are coupled, and I'm looking for a way to properly incorporate these conditions for the outlet. I am wondering how I can implement these by MOOSE. Is there any availabe BCs or should I write my own?

Thanks a lot. 

Bests, 
Mehdi

**Comment by GiudGiud** [comment]
Hello

I think you could have a Dirichlet boundary condition for p1 that is simply: p1 = p2 J1 / J2 with p1 being the boundary condition variable and all the others being coupled variables. 

Are J1 and J2 nonlinear variables? (in the system, being solved for) or are they postprocessed quantities?
I have not seen this anywhere else, you will probably have to write your own.

Guillaume

**Reply by Meghasemi-96** [reply]
Hello Guillaume,

Thanks for your response. The governing equation for my case is J = -D(dp/dx), where J represents flux, D stands for diffusion, and p refers to pressure. The solvation is in a steady state. I am dealing with a binary gas (single phase, two components) transport process in a 2D structure along the x-direction.

I don't encounter any challenges with the inlet boundary condition since the values of pressure for both components are determined. However, at the outlet, I need to apply two boundary conditions where the exact values of the pressure for each component are not determined. Instead, I must satisfy the conditions `p1 + p2 = 1` and `J1/J2 = p1/p2`. It appears that one of these conditions should be applied as a constraint, but I'm uncertain about the appropriate way to implement them. 

As you mentioned, J is a post-processed quantity. In each step of the solvation process, first, values should be assigned to `p1` and `p2` to accurately satisfy the condition `p1 + p2 = 1`. Afterward, J should be calculated as a post-processed quantity to verify whether the chosen `p1 `and `p2` values for the outlet meet this condition `J1/J2 = p1/p2`. If the condition `J1/J2 = p1/p2` is not fulfilled, it is needed to make new guesses for `p1` or `p2`. An iterative solvation approach should be employed, where the values of` p1` and `p2` should be adjusted until the condition `J1/J2 = p1/p2` is satisfied.


Thanks. 


Bests, 
Mehdi

**Reply by GiudGiud** [reply]
Could you compute J in the boundary conditions object?
That would make things simple, you could impose p1+p2=1 in the p1 equation and J1/J2 = p1/p2 in the boundary condition for p2

both could be Dirichlet boundary conditions

**Reply by maxnezdyur** [reply]
I am unsure if I understand what you need to do, but after solving the system of equations for your BCs, you could apply the BCs below for p1 and p2, and you wouldn't have to use an iterative approach.
1. $p_1 = \frac{J_1}{J_1 + J_2} $
2. $p_2 = \frac{J_2}{J_1 + J_2} $

**Comment by Meghasemi-96** [comment]
Hello,
 
Thanks for your help. I followed this idea by calculating the flux (J) with a postprocessor (SideDiffusiveFluxAverage) and imposing the values for each variable on boundary with CoupledDirichletBC, and it worked. However, this only works if the diffusivity values are constant numbers (`J = -D(dp/dx)`, and D=const). If I make diffusivity values a function of variables, I kind of face a mass balance problem and the input and output flux are not the same anymore (they are for constant diffusivity case).
In more details, I want to simulate binary gas transport where only Fickian diffusion (`J = -D(dp/dx)`) exists. Here, for each gas, diffusion is a function of its partial pressure: `D1 = f(p1)`  and `D2= f(p2)` . The BC of inlet is `p1=p2=5` and the outlet BC is `p1+p2=1` and `J1/J2 = p1/p2`. The problem is solved but at the end the fluxes are not the same in the inlet and outlet for each variable as it should be like the constant diffusivity case (see below).


| time                  | uoutppl           | uoutppr                | voutppl        | voutppr        |
+----------------+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|  1.000000e+00  |  -6.600876e-01 |   6.600876e-01   |  -3.754483e-01 |   3.754483e-01 |

Which uoutppl, uoutppr, voutppl, voutppr are left and right boundary postprocessor (fluxes) values for variables u and v, respectively. So I was wondering if you could help me understand how this discrepancy happens in variable dependent case. I set the execute_on parameter for postprocessor to after each linear iteration as well, to keep it update. Please let me know your thoughts. Thanks.

**Reply by GiudGiud** [reply]
Hello

u = p1 and v=p2?

So you are imposing `J1/J2 = p1/p2` but the postprocessing is not showing this is working?
p1=p2 on the inlet is working and shown to work by a SideAveragePP right?

Guillaume

**Reply by Meghasemi-96** [reply]
Dear Guillaume,
Thanks for your reply.
 
Yes u and v are pressure variables p1 and p2. As I mentioned we imposed the two boundary conditions with CoupledDirichletBC and using postprocessors. The first condition is `p1+p2=1` which is implemented as follow:
 
```
#include "CoupledDirichletBC.h"
registerMooseObject("MOOSE_MMMApp", CoupledDirichletBC);
 
InputParameters
CoupledDirichletBC::validParams()
{
  InputParameters params = NodalBC::validParams();
 
  // Specify input parameters that we want users to be able to set:
  params.addParam<Real>("alpha", 1.0, "Value multiplied by the coupled value on the boundary");
  params.addRequiredCoupledVar("some_var", "Value on the boundary");
  return params;
}
 
CoupledDirichletBC::CoupledDirichletBC(const InputParameters & parameters)
  : NodalBC(parameters),
    // store the user-specified parameters from the input file...
    _alpha(getParam<Real>("alpha")),        // for the multiplier
    _some_var_val(coupledValue("some_var")) // for the coupled variable
{
}
 
Real
CoupledDirichletBC::computeQpResidual()
{
  // For dirichlet BCS, u=BC at the boundary, so the residual includes _u and the desired BC value:
  return _u[_qp] - (_alpha - _some_var_val[_qp]);
}
```
and is called in the input file like this:
```
    [r3]
      type = CoupledDirichletBC
      variable = u
      boundary = 'front'
      some_var = v
      alpha = 1
    []
```
 
The other condition is `J1/J2 = p1/p2` which is written and used as follows:
 
```
#include "Mehdi.h"
 
registerMooseObject("MOOSE_MMMApp", Mehdi);
 
InputParameters
Mehdi::validParams()
{
  InputParameters params = NodalBC::validParams();
 
  // Specify input parameters that we want users to be able to set:
  // params.addParam<Real>("alpha", 1.0, "Value multiplied by the coupled value on the boundary");
  params.addRequiredCoupledVar("some_var", "Value on the boundary");
  params.addRequiredParam<PostprocessorName>(
      "postprocessor1", "The postprocessor to set the value to on the boundary.");
  params.addRequiredParam<PostprocessorName>(
      "postprocessor2", "The postprocessor to set the value to on the boundary.");
  return params;
}
 
Mehdi::Mehdi(const InputParameters & parameters)
  : NodalBC(parameters),
    // store the user-specified parameters from the input file...
    // _alpha(getParam<Real>("alpha")),        // for the multiplier
    _some_var_val(coupledValue("some_var")), // for the coupled variable
    _postprocessor_value1(getPostprocessorValue("postprocessor1")),
    _postprocessor_value2(getPostprocessorValue("postprocessor2"))
{
}
 
Real
Mehdi::computeQpResidual()
{
  // For dirichlet BCS, u=BC at the boundary, so the residual includes _u and the desired BC value:
  if (_postprocessor_value2==0)
  {
  return _u[_qp] - ((_postprocessor_value1) * _some_var_val[_qp]);
  }
  else
  return _u[_qp] - ((_postprocessor_value1/_postprocessor_value2) * _some_var_val[_qp]);
}
```
```
    [r4]
      type = Mehdi
      variable = v
      boundary = 'front'
      some_var = u
      postprocessor1 = voutppr
      postprocessor2 = uoutppr
    []
```
and hence the postprocessors voutppr and uoutppr for right (front in 3D) boundary voutppl and uoutppl for left (back in 3D) boundary are defined as:
```
  [Postprocessors]
    [uoutppr]
      type = SideFluxAverage
      variable = u
      boundary = front
      diffusivity = D1
      execute_on = 'INITIAL LINEAR'
      outputs = console
    []
    [voutppr]
      type = SideFluxAverage
      variable = v
      boundary = front
      diffusivity = D2
      execute_on = 'INITIAL LINEAR'
      outputs = console
    []
    [uoutppl]
      type = SideFluxAverage
      variable = u
      boundary = back
      diffusivity = D1
      outputs = console
      execute_on = 'INITIAL LINEAR'
    []
    [voutppl]
      type = SideFluxAverage
      variable = v
      boundary = back
      diffusivity = D2
      outputs = console
      execute_on = 'INITIAL LINEAR'
    []
  []
```
 
The code has been benchmarked in constant diffusivity cases and as showed in the previous comment the flux values are equal for each variable in two boundaries:
```
  [Materials]
    [diffu0]
      type = ParsedMaterial
      property_name = D1
    expression = '1e+5'
    coupled_variables = 'u v'
    block = 0
    []
    [diffv0]
        type = ParsedMaterial
        property_name = D2
        expression = '1.667e+3'
        coupled_variables = 'u v'
        block = 0
    []
    [diffu1]
      type = ParsedMaterial
      property_name = D1
      expression = '1e+3'
      coupled_variables = 'u v'
      block = 1
    []
    [diffv1]
        type = ParsedMaterial
        property_name = D2
        expression = '6.667'
        coupled_variables = 'u v'
        block = 1
    []
  []
```
```
| time | uoutppl | uoutppr | voutppl | voutppr |
+----------------+----------------+----------------+----------------+----------------+
| 0.000000e+00 | 0.000000e+00 | 0.000000e+00 | 0.000000e+00 | 0.000000e+00 |
| 1.000000e+00 | -6.600876e-01 | 6.600876e-01 | -3.754483e-01 | 3.754483e-01 |
```
 
But as diffusivity becomes a function of variables u and v, then the problem is solved but flux values are not equal for each variable which is not correct. So I was wondering where the issue comes from to solve it, is it because of using postprocessors or something else. Please let me know if you need further information. Thanks.

**Reply by GiudGiud** [reply]
so these two constraints are imposed both with nodal BCs on the same boundary?

Using postprocessors should not affect the result. It will slow down convergence as contributions to the Jacobian are lost when the numbers are passed through a postprocessor. 

The nodal bcs/constraints are being imposed on every node. Should this entail that the integration over all nodes on the boundaries has this equality as well?
What I mean is that imposing a ratio of J1/J2 = p1/p2 on every node will get:

integral(J1/J2) = integral(p1/p2)

it wont get

integral(J1) = intgral(J2) * Integral(p1/p2)

**Reply by Meghasemi-96** [reply]
Dear Guillaume,
Thanks for your reply.
 
Yes, you are correct. I believe we are looking for the first scenario where integral (J1/J2) = integral(p1/p2), however, as I cannot think of other reason for this problem. I can check the other way as well. Would you please advise how I can implement the second boundary condition where integral(J1) = integral(J2) * Integral(p1/p2)?

**Reply by GiudGiud** [reply]
I think you may already be in the second case, since you compute J1 and J2 in a global way then impose it in a local manner to p1 and p2.
This was just an example. my intuition is that something similar is happening when D_1/2 are not constant

**Reply by Meghasemi-96** [reply]
Dear Guillaume,

Thanks for your prompt response. I see. I was wondering how this constraint can be imposed locally? Should J1 and J2 be calculated directly in BCs locally instead of using postprocessors?

**Reply by GiudGiud** [reply]
You could try that but it will likely not be straightforward inside a nodal bc to compute the current which is integrated on sides instead

You could store the diffusive current in an auxiliary variable and use that to get the local value of the current
